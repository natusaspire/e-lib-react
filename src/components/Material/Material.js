import React, { PureComponent } from 'react';

import {
  TEXT_MATERIAL,
  AUDIO_MATERIAL,
  VIDEO_MATERIAL
} from '@/constants/materialTypes';
import MaterialTypeWrapper from '@/components/MaterialTypeWrapper';
import MaterialText from '@/components/MaterialText';
import MaterialAudio from '@/components/MaterialAudio';
import MaterialVideo from '@/components/MaterialVideo';
import Error from '@/components/Error';

const renderMaterialTypeComponent = material => {
  switch (material.type) {
    case TEXT_MATERIAL:
      return <MaterialText material={material} />;
    case AUDIO_MATERIAL:
      return <MaterialAudio material={material} />;
    case VIDEO_MATERIAL:
      return <MaterialVideo material={material} />;
    default:
      return null;
  }
};

class Material extends PureComponent {
  componentDidMount() {
    const {
      getMaterial,
      match: { params }
    } = this.props;

    getMaterial(params);
  }

  componentDidUpdate() {
    const {
      material: { data }
    } = this.props;

    if (data) {
      document.title = data.title;
    }
  }

  componentWillUnmount() {
    const { unsetMaterial } = this.props;

    unsetMaterial();
  }

  render() {
    const {
      material: { data, error }
    } = this.props;

    return error ? (
      <Error status={error} />
    ) : (
      <MaterialTypeWrapper>
        {!!data && renderMaterialTypeComponent(data)}
      </MaterialTypeWrapper>
    );
  }
}

export default Material;
