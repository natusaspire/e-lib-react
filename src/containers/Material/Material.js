import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { getMaterial, unsetMaterial } from '@/actions/material';
import Material from '@/components/Material';

const mapStateToProps = ({ material }) => ({ material });

const mapDispatchToProps = dispatch =>
  bindActionCreators({ getMaterial, unsetMaterial }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Material);
