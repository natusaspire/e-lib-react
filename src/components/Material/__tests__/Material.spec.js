import React from 'react';
import { shallow } from 'enzyme';

import Material from '@/components/Material';

jest.mock('@/constants/materialTypes', () => ({
  TEXT_MATERIAL: 'text',
  AUDIO_MATERIAL: 'audio',
  VIDEO_MATERIAL: 'video'
}));

jest.mock('@/components/MaterialTypeWrapper', () => 'MaterialTypeWrapper');
jest.mock('@/components/MaterialText', () => 'MaterialText');
jest.mock('@/components/MaterialAudio', () => 'MaterialAudio');
jest.mock('@/components/MaterialVideo', () => 'MaterialVideo');
jest.mock('@/components/Error', () => 'Error');

describe('<Material />', () => {
  const props = {
    match: {
      params: {
        endpoint: 'audio',
        id: '1'
      }
    },
    material: {
      loading: false,
      error: null,
      data: {
        id: 1,
        type: 'audio',
        title: 'Material title',
        url: 'audio.mp3'
      }
    },
    getMaterial: jest.fn(),
    unsetMaterial: jest.fn()
  };

  beforeEach(() => {
    global.document.title = '';
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render text material when data exists and data type is text', () => {
    const material = {
      data: {
        id: 1,
        type: 'text',
        title: 'Material title',
        textContent: 'Material text content.'
      },
      loading: false,
      error: null
    };

    const wrapper = shallow(<Material {...props} material={material} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render audio material when data exists and data type is audio', () => {
    const material = {
      data: {
        id: 1,
        type: 'audio',
        title: 'Material title',
        url: 'audio.mp3'
      },
      loading: false,
      error: null
    };

    const wrapper = shallow(<Material {...props} material={material} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render video material when data exists and data type is video', () => {
    const material = {
      data: {
        id: 1,
        type: 'video',
        title: 'Material title',
        url: 'video.mp4'
      },
      loading: false,
      error: null
    };

    const wrapper = shallow(<Material {...props} material={material} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render material type wrapper when data exists and data type is unknown', () => {
    const material = {
      data: {
        id: 1,
        type: 'type',
        title: 'Material title',
        url: 'video.mp4'
      },
      loading: false,
      error: null
    };

    const wrapper = shallow(<Material {...props} material={material} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render material type wrapper when data does not exist', () => {
    const material = {
      data: null,
      loading: true,
      error: null
    };

    const wrapper = shallow(<Material {...props} material={material} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render error content when error', () => {
    const material = {
      data: null,
      loading: false,
      error: '404'
    };

    const wrapper = shallow(<Material {...props} material={material} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should get material on mount', () => {
    shallow(<Material {...props} />);

    expect(props.getMaterial).toHaveBeenCalledWith(props.match.params);
  });

  it('should change document title on update when material data exists', () => {
    const material = {
      ...props.material,
      data: null
    };

    const wrapper = shallow(<Material {...props} material={material} />);

    wrapper.setProps({ material: props.material });

    expect(global.document.title).toBe(props.material.data.title);
  });

  it('should not change document title on update when material data does not exist', () => {
    const material = {
      ...props.material,
      data: null
    };

    const wrapper = shallow(<Material {...props} material={material} />);

    wrapper.setProps({
      material: { ...material }
    });

    expect(global.document.title).toBe('');
  });

  it('should unset material on unmount', () => {
    const wrapper = shallow(<Material {...props} />);

    wrapper.instance().componentWillUnmount();

    expect(props.unsetMaterial).toHaveBeenCalled();
  });
});
