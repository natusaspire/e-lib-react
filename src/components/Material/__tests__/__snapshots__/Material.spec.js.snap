// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Material /> should render audio material when data exists and data type is audio 1`] = `
<MaterialTypeWrapper>
  <MaterialAudio
    material={
      Object {
        "id": 1,
        "title": "Material title",
        "type": "audio",
        "url": "audio.mp3",
      }
    }
  />
</MaterialTypeWrapper>
`;

exports[`<Material /> should render error content when error 1`] = `
<Error
  status="404"
/>
`;

exports[`<Material /> should render material type wrapper when data does not exist 1`] = `<MaterialTypeWrapper />`;

exports[`<Material /> should render material type wrapper when data exists and data type is unknown 1`] = `<MaterialTypeWrapper />`;

exports[`<Material /> should render text material when data exists and data type is text 1`] = `
<MaterialTypeWrapper>
  <MaterialText
    material={
      Object {
        "id": 1,
        "textContent": "Material text content.",
        "title": "Material title",
        "type": "text",
      }
    }
  />
</MaterialTypeWrapper>
`;

exports[`<Material /> should render video material when data exists and data type is video 1`] = `
<MaterialTypeWrapper>
  <MaterialVideo
    material={
      Object {
        "id": 1,
        "title": "Material title",
        "type": "video",
        "url": "video.mp4",
      }
    }
  />
</MaterialTypeWrapper>
`;
